core/libp2p/muxer/yamux/yamux.cpp

```diff
@@ -0,0 +1,521 @@
+      return "the stream is closed for reads";
+    case ErrorType::YAMUX_IS_CLOSED:
+      return "this Yamux instance is closed";
+  }
+}
+
+namespace libp2p::muxer {
+  Yamux::Yamux(std::shared_ptr<transport::Connection> connection,
+               NewStreamHandler stream_handler, YamuxConfig yamux_config,
+               kagome::common::Logger logger)
+      : connection_{std::move(connection)},
+        new_stream_handler_{std::move(stream_handler)},
+        is_active_{true},
+        config_{yamux_config},
+        logger_{std::move(logger)} {
+    // client uses odd numbers, server - even
+    last_created_stream_id_ = config_.is_server ? 0 : 1;
+  }
+
+  Yamux::~Yamux() = default;
+
+  void Yamux::start() {
+    startReadingHeader();
+  }
+
+  void Yamux::startReadingHeader() {
+    if (!is_active_) {
+      return;
+    }
+    if (connection_->isClosed()) {
+      closeYamux();
+      return;
+    }
+
+    connection_->asyncRead(
+        read_buffer_, YamuxFrame::kHeaderLength,
+        [t = shared_from_this()](const std::error_code &ec, size_t n) {
+          t->readingHeaderCompleted(ec, n);
+        });
+  }
+
+  void Yamux::readingHeaderCompleted(const std::error_code &ec, size_t n) {
+    if (ec) {
+      logger_->error("cannot read from from the connection: {}", ec.value());
+      // terminate the Yamux? It's a critical error
+    }
+    if (n != YamuxFrame::kHeaderLength) {
+      logger_->error(
+          "connection error: read less bytes, than expected in header");
+      // terminate the Yamux?
+    }
+
+    if (!processHeader()) {
+      startReadingHeader();
+    }
+  }
+
+  void Yamux::readingDataCompleted(const std::error_code &ec, size_t n,
+                                   StreamParameters &stream) {
+    if (ec) {
+      logger_->error("cannot read from the connection: {}", ec.value());
+      // terminate Yamux?
+    }
+
+    // if there's a callback, which awaits for the message from this stream,
+    // call it; buffer the message otherwise
+    std::vector<uint8_t> bytes(n);
+    boost::asio::buffer_copy(boost::asio::buffer(bytes), read_buffer_.data());
+    read_buffer_.consume(n);
+    kagome::common::Buffer msg{std::move(bytes)};
+    if (!stream.completion_handlers_.empty()) {
+      stream.completion_handlers_.front()(std::move(msg));
+      stream.completion_handlers_.pop();
+    } else {
+      stream.buffered_messages_.push(std::move(msg));
+    }
+
+    startReadingHeader();
+  }
+
+  void Yamux::write(const common::NetworkMessage &msg,
+                    stream::Stream::ErrorCodeCallback cb) {
+    outcoming_messages_.push({msg, std::move(cb)});
+    if (!is_writing_) {
+      startWriting();
+    }
+  }
+
+  void Yamux::startWriting() {
+    if (!is_active_) {
+      return;
+    }
+    if (connection_->isClosed()) {
+      closeYamux();
+      return;
+    }
+    is_writing_ = true;
+
+    if (!outcoming_messages_.empty()) {
+      auto msg_and_callback = outcoming_messages_.front();
```

Comment:
no need to move buffers around, just don't pop it here, and pop it after msg has beer written/read